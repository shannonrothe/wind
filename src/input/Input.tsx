import React from "react";
import { BaseProps } from "../types";

export type InputProps = {
  onChange?: (event: React.ChangeEvent) => void;
  disabled?: boolean;
  label?: string;
  placeholder?: string;
  type?: "text" | "password" | "number";
  value?: string;
} & BaseProps;

type InputClassList = (string | false | undefined)[];

export const Input = (props: InputProps) => {
  const {
    onChange,
    disabled,
    id,
    label,
    placeholder,
    type = "text",
    value,
  } = props;
  const classes: InputClassList = [
    "appearance-none",
    "w-full",
    "px-4 py-2",
    disabled ? "bg-gray-200" : "bg-white",
    "border",
    disabled && "border-gray-300",
    "rounded",
    "text-sm",
    "focus:outline-none focus:shadow-outline focus:border-none",
    "text-gray-800",
  ];

  // Extracted to a variable so it can be reused between label and no label
  const InputElement = () => (
    <input
      id={id}
      disabled={disabled}
      value={value}
      onChange={onChange}
      placeholder={placeholder}
      className={`${classes.join(" ")}`}
      type={type}
    />
  );

  return (
    <div>
      {label && (
        <label htmlFor={id}>
          <span className="inline-block mb-1 text-xs font-semibold text-gray-600 uppercase">
            {label}
          </span>
          <InputElement />
        </label>
      )}

      {!label && <InputElement />}
    </div>
  );
};
