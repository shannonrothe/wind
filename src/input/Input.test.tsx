import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Input, InputProps } from "./Input";
import { mount } from "enzyme";

describe("Input", () => {
  let defaultProps: InputProps = {};

  beforeEach(() => {
    defaultProps.onChange = jest.fn();
    defaultProps.value = "";
  });

  it("renders", () => {
    expect(render(<Input {...defaultProps} />)).toMatchSnapshot();
  });

  it("renders with a label", () => {
    expect(render(<Input {...defaultProps} label='Input' />)).toMatchSnapshot();
  });

  it("is disabled when disabled prop is true", () => {
    const wrapper = mount(<Input {...defaultProps} disabled={true} />);

    expect(wrapper.find("input").getDOMNode().hasAttribute("disabled")).toBe(true);
  });

  it("triggers onChange when not disabled", () => {
    const wrapper = mount(<Input {...defaultProps} />);
    const event = {
      preventDefault() {},
      target: { value: "the-value" },
    };

    wrapper.find("input").simulate("change", event);
    expect(defaultProps.onChange).toHaveBeenCalled();
  });

  it("doesn't trigger onChange when disabled", () => {
    const wrapper = mount(<Input {...defaultProps} disabled={true} />);
    const event = {
      preventDefault() {},
      target: { value: "the-value" },
    };

    wrapper.find("input").simulate("keypress", event);
    expect(defaultProps.onChange).not.toHaveBeenCalled();
  });
});
