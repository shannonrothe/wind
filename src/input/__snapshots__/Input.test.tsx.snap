// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <input
          class="appearance-none w-full px-4 py-2 bg-white border  rounded text-sm focus:outline-none focus:shadow-outline focus:border-none text-gray-800"
          type="text"
          value=""
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <input
        class="appearance-none w-full px-4 py-2 bg-white border  rounded text-sm focus:outline-none focus:shadow-outline focus:border-none text-gray-800"
        type="text"
        value=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Input renders with a label 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <label>
          <span
            class="inline-block mb-1 font-semibold text-gray-600 uppercase text-xs"
          >
            Input
          </span>
          <input
            class="appearance-none w-full px-4 py-2 bg-white border  rounded text-sm focus:outline-none focus:shadow-outline focus:border-none text-gray-800"
            type="text"
            value=""
          />
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <label>
        <span
          class="inline-block mb-1 font-semibold text-gray-600 uppercase text-xs"
        >
          Input
        </span>
        <input
          class="appearance-none w-full px-4 py-2 bg-white border  rounded text-sm focus:outline-none focus:shadow-outline focus:border-none text-gray-800"
          type="text"
          value=""
        />
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
